src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,152 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
+    public ArrayList(int capacity) {
+        elementData = new Object[capacity];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        if (isFull()) {
+            elementData = grow();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexForAdd(index);
+        if (isFull()) {
+            elementData = grow();
+        }
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int numberOfValuesToAdd = list.size();
+        Object[] valuesToAdd = convertListToArray(list);
+        if (numberOfValuesToAdd == 0) {
+            return;
+        }
+        if (numberOfValuesToAdd > elementData.length - size) {
+            elementData = grow(size + numberOfValuesToAdd);
+            System.arraycopy(valuesToAdd, 0, elementData, size, numberOfValuesToAdd);
+        }
+        size += numberOfValuesToAdd;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T oldElement = get(index);
+        fastRemove(index);
+        return oldElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int indexOfElement = findElement(element);
+        return remove(indexOfElement);
+    }
+
+    private void fastRemove(int index) {
```

Comment:
Probably, `internalRemove` will be more suitable here, what do you think? `fastRemove` makes me feel it is somehow optimised.

Reply:
After you told me, I felt the same way. In case we have the "remove" method, the "internalRemove" name will be the best neutral renaming variant. Fixed.